modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/CompositePublisher.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.sql.engine.util;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.Flow;
+import java.util.concurrent.Flow.Publisher;
+import java.util.concurrent.Flow.Subscriber;
+import java.util.concurrent.Flow.Subscription;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.jetbrains.annotations.Nullable;
+
+public class CompositePublisher<T> implements Flow.Publisher<T> {
+    private final Collection<Publisher<T>> publishers = new ArrayList<>();
+
+    private final SubscriptionManagementStrategy<T> subscriptionStrategy;
+
+    private final AtomicBoolean subscribed = new AtomicBoolean();
+
+    private final boolean ordered;
+
+    public CompositePublisher(@Nullable Comparator<T> comp) {
+        ordered = comp != null;
+
+        if (ordered)
+            subscriptionStrategy = new OrderedInputSubscriptionStrategy<>(comp);
+        else
+            subscriptionStrategy = new PlainSubscriptionManagementStrategy<>();
+    }
+
+    public void add(Publisher<T> publisher) {
+        if (subscribed.get())
+            throw new IllegalStateException("Cannot add publisher after subscription.");
+
+        publishers.add(publisher);
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> delegate) {
+        if (!subscribed.compareAndSet(false, true)) {
+            throw new IllegalStateException("Multiple subscribers are not supported.");
+        }
+
+        int idx = 0;
+
+        for (Publisher<T> publisher : publishers) {
+            Subscriber<T> subscriber = wrap((Subscriber<T>) delegate, idx++);
+
+            subscriptionStrategy.addSubscriber(subscriber);
```

Comment:
why `addSubscriber`? If we can have just single value `set` looks more preferable